server/keyspace/keyspace.go

```diff
@@ -0,0 +1,318 @@
+// Copyright 2022 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package keyspace
+
+import (
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/keyspacepb"
+	"github.com/tikv/pd/pkg/syncutil"
+	"github.com/tikv/pd/server/id"
+	"github.com/tikv/pd/server/storage/endpoint"
+)
+
+const (
+	// AllocStep set idAllocator's step when write persistent window boundary.
+	// Use a lower value for denser idAllocation in the event of frequent pd leader change.
+	AllocStep = uint64(100)
+	// AllocLabel is used to label keyspace idAllocator's metrics.
+	AllocLabel = "keyspace-idAlloc"
+	// DefaultKeyspaceName is the name reserved for default keyspace.
+	DefaultKeyspaceName = "DEFAULT"
+	// DefaultKeyspaceID is the id of default keyspace.
+	DefaultKeyspaceID = uint32(0)
+)
+
+// Manager manages keyspace related data.
+// It validates requests and provides concurrency control.
+type Manager struct {
+	// idLock guards keyspace name to id lookup entries.
+	idLock syncutil.Mutex
+	// metaLock guards keyspace meta.
+	metaLock *lockGroup
+	// idAllocator allocates keyspace id.
+	idAllocator id.Allocator
+	// store is the storage for keyspace related information.
+	store endpoint.KeyspaceStorage
+}
+
+// CreateKeyspaceRequest represents necessary arguments to create a keyspace.
+type CreateKeyspaceRequest struct {
+	// Name of the keyspace to be created.
+	// Using an existing name will result in error.
+	Name   string
+	Config map[string]string
+	Now    time.Time
+}
+
+// NewKeyspaceManager creates a Manager of keyspace related data.
+func NewKeyspaceManager(store endpoint.KeyspaceStorage, idAllocator id.Allocator) (*Manager, error) {
+	manager := &Manager{
+		store:       store,
+		idAllocator: idAllocator,
+		metaLock:    newLockGroup(),
+	}
+
+	// Check if default keyspace already exists.
+	defaultExists, err := manager.store.LoadKeyspace(DefaultKeyspaceID, &keyspacepb.KeyspaceMeta{})
+	if err != nil {
+		return nil, err
+	}
+	if defaultExists {
+		return manager, nil
+	}
+	// Initialize default keyspace.
+	defaultKeyspace := &keyspacepb.KeyspaceMeta{
```

Comment:
Why not use CreateKeyspace to init the default keyspace? Seems it has more information like creating time and so on.